### DISCUSSION TITLE
Question on stochastic module LibtorchANNTrainer example input file GFunction

### DISCUSSION THREAD

**Discussion by chongxiyuan** [idaholab/moose#24368] [Q&A Modules: General] [answered]
Hi MOOSE community,

I recently installed MOOSE with libtorch, and I am trying to learn how to train and evaluate neural networks on MOOSE. I just started running given examples. I run into an error for modules/stochastic_tools/test/tests/surrogates/libtorch_nn/train.i file that says: `A 'GFunction' is not a registered object.` 

It seems like GFunction is a vectorpostprocessor, but I didn't find it in the MOOSE website documentation. Is this a custom postprocessor? Please point me to the right direction.

The example input file is 

```
[StochasticTools]

[]

[Samplers]
  [sample]
    type = CartesianProduct
    linear_space_items = '0 0.0125 5
                          0 0.0125 5
                          0 0.0125 5'
  []
[]

[VectorPostprocessors]
  [values]
    type = GFunction
    sampler = sample
    q_vector = '0 0 0'
    execute_on = INITIAL
    outputs = none
  []
[]

[Trainers]
  [train]
    type = LibtorchANNTrainer
    sampler = sample
    response = values/g_values
    num_epochs = 40
    num_batches = 10
    num_neurons_per_layer = '64 32'
    learning_rate = 0.001
    nn_filename = mynet.pt
    read_from_file = false
    print_epoch_loss = 10
    activation_function = 'relu relu'
    max_processes = 1
  []
[]

[Outputs]
  [out]
    type = SurrogateTrainerOutput
    trainers = 'train'
    execute_on = FINAL
  []
[]
```

Thank you!

Chongxi Yuan

**Comment by GiudGiud** [comment, marked as answer]
Hello

This is a test object for the stochastic tools module.
If you want to use it, you need to
- use the stochastic_tools module executable
- pass on the command line: --allow-test-objects 

It wont have documentation as it s a test object. But you can find it in modules/stochastic_tools/test/src/functions

Guillaume

**Reply by GiudGiud** [reply]
@zachmprince fyi in case it needs to be promoted

**Reply by chongxiyuan** [reply]
Thank you so much, I found the function in modules/stochastic_tools/test/src/functions and manually copied the function to my executable folder, and it worked as well.

**Reply by zachmprince** [reply]
`GFunction` is something commonly used for verification of global sensitivity methods. I don't think it has any physical meaning. So I think having it as a test-only object is appropriate. You can probably replicate it using a `ParsedPostprocessor` in a sub-application.
